src/pudl/output/ferc1.py

```diff
@@ -1063,6 +1077,88 @@ def calculations(self):
         ].copy()
         return calc_explode
 
+    @property
+    def calculations(self: Self):
+        """Remove any calculation components that aren't relevant to the explosion.
+
+        At the end of this process several things should be true:
+
+        - Only parents with table_name in the explosion tables should be retained.
+        - All calculations in which *any* components were outside of the tables in
+          the explosion should be turned into leaves -- i.e. they should be replaced
+          with a single calculation component filled with NA values.
+        - Every remaining calculation component must also appear as a parent (if it
+          is a leaf, then it will have a single null calculation component)
+        - There should be no records where only one of table_name or xbrl_factoid
+          are null. They should either both or neither be null.
+        - table_name_parent and xbrl_factoid_parent should be non-null.
+        """
+        calc_cols = list(NodeId._fields)
+        parent_cols = [col + "_parent" for col in calc_cols]
+
+        # Keep only records where parent is within exploded tables:
+        calc_explode = self.calculation_components_xbrl_ferc1[
+            self.calculation_components_xbrl_ferc1.table_name_parent.isin(
+                self.table_names
+            )
+        ].copy()
+        # Groupby parent factoids
+        gb = calc_explode.groupby(
+            ["table_name_parent", "xbrl_factoid_parent"], as_index=False
+        )
+        # Identify groups where all calculation components are within the explosion
+        # We need to include table_name of pd.NA to retain non-calculated values.
+        calc_explode["is_within_explosion"] = gb.table_name.transform(
+            lambda x: x.isin(self.table_names + [pd.NA]).all()
+        )
+        # Convert any calculation componets from outside the explosion into reported
+        # values / leaf nodes:
+        calc_explode.loc[
+            ~calc_explode.is_within_explosion, list(NodeId._fields) + ["weight"]
+        ] = pd.NA
+        calc_explode = calc_explode.drop(columns=["is_within_explosion"]).reset_index(
+            drop=True
+        )
```

Comment:
The intention with:
```py
        calc_explode.loc[
            ~calc_explode.is_within_explosion, list(NodeId._fields) + ["weight"]
        ] = pd.NA
```
is to set all of the calculation component fields to NA, which turns their parents into leaves. But I guess then there are many identical duplicate null records, since there's one for every former component, and we just want a single one for the parent (with no other values in the record).